Use search indices

dranikpg opened the pull request:
Context:

Currently querying is done with brute force search. Indices for specific fields should allow querying more efficiently.

Goal:
* Support descriptor for schemas
* Support basic indices for numbers & text
* Support basic search algorithm to merge the results (no optimizations)

This is the step after #1249. Once its merged, we'll be able to plug the integers ids in here into the external interface

dranikpg commented:
The goods news: I've replaced the whole pipeline to use indexed search 🥳 
The bad news: the changediff 🙄

dranikpg commented:
Now its much smaller and can be reviewed

romange commented on src/core/search/search.cc:
```diff
@@ -24,57 +25,154 @@ AstExpr ParseQuery(std::string_view query) {
+    static_assert(is_base_of_v<BaseIndex, T>);
+    auto index = indices->GetIndex(field);
+    DCHECK(index);  // TODO: handle not existing erorr
+    auto* casted_ptr = dynamic_cast<T*>(index);
```
dynamic_cast is cheating :)

dranikpg replied:
It can be of multiple types 💁🏻There are two options, variants and virtual pointers. 

Virtual casting is a perfectly fine and safe cast.

romange replied:
it uses RTTI, no?
also, it throws exceptions that we do not support.

romange replied:
anyway, it's fine, it's just my personal taste.

dranikpg replied:
Yes, it uses RTTI. Like every virtual method call. 

It throws only if the pointer is invalid. If the cast failed, it returns nullptr.

romange replied:
it's not accurate. rtti is special. See https://gcc.gnu.org/onlinedocs/gcc-4.4.7/gcc/C_002b_002b-Dialect-Options.html

```
Disable generation of information about every class with virtual functions for use by the C++ runtime type identification features (`dynamic_cast' and `typeid'). If you don't use those parts of the language, you can save some space by using this flag.
```

chakaz replied:
Passer-by comment:

RTTI is somewhat different (think of it as an extension) to virtual calls.
It's true that virtual calls use a vtable, with a (small) cost of double dereference for invoking methods. However RTTI requires additional information (which can be disabled when not affecting virtual calls).
Disabling RTTI can save in binary size (I can't estimate but I would imagine it's negligible at least in our case). Also RTTI casts (i.e. casting down using `dynamic_cast`) is slower than regular casts or even virtual calls.

Again, I don't think it should not be used :)

There's some more info here for the curious: https://shaharmike.com/cpp/vtable-part1/

romange commented on src/core/search/search.h:
```diff
@@ -4,12 +4,51 @@
+
+  void Add(DocId doc, DocumentAccessor* access);
+
+  BaseIndex* GetIndex(std::string_view field);
```
why not const?

romange commented on src/core/search/search.h:
```diff
@@ -4,12 +4,51 @@
+};
+
+// Collection of indices for all fields in schema
+struct FieldIndices {
```
class

romange commented on src/server/search/doc_index.cc:
```diff
@@ -84,21 +84,29 @@ uint8_t DocIndex::GetObjCode() const {
+  for (search::DocId doc : doc_ids) {
+    auto key = key_index_.Get(doc);
+    auto it = db_slice.Find(op_args.db_cntx, key, base_->GetObjCode());
+    CHECK(it);
```
when it will fail you will want to know the key :)

dranikpg commented:
Fixed const-ness. Added it more on all levels, though it'll need to be reverted on some for the dynamic updates.

Didn't understand why using dynamic_cast is cheating by default. Indices can be of multiple types - there are two options: variants and virtuality. I chose the second. They need to by type-asserted when read.

(Well the third option would be storing a separate map for every type, as there will be only a few, but...)

romange commented on src/core/search/search_test.cc:
```diff
@@ -94,15 +79,24 @@ class SearchParserTest : public ::testing::Test {
-    for (auto& [doc, expect_match] : entries_) {
-      bool doc_matched = search_algo.Check(&doc);
+    if (!is_sorted(matched.begin(), matched.end()))
+      throw "Search result is not sorted";
```
LOG(FATAL) ?

romange commented on src/server/search/doc_accessors.cpp:
```diff
@@ -28,39 +28,20 @@ string_view SdsToSafeSv(sds str) {
 
 }  // namespace
 
-bool ListPackAccessor::Check(FieldConsumer f, string_view active_field) const {
```
we use cc extension :)

dranikpg replied:
Good catch 😅
